### PR TITLE
Add PIL formats(gif,png...) support for retrain.py

### PR DESCRIPTION
**Pull request by elviswf**
- Add PIL formats(gif,png...) support for retrain.py.
- Use PIL.Image to read images and io.BytesIO to serialize them.

### CODE REVIEW

**Comment by elviswf** [comment]
**I signed it!**
Thank you!

File: tensorflow/examples/image_retraining/retrain.py [hunk 1]
```diff
@@ -73,6 +73,8 @@
 import sys
 import tarfile
 
+from PIL import Image
```
Comment (rohan100jain) [inline tensorflow/examples/image_retraining/retrain.py]: Might want to add in the comments above that PIL would need to be installed in order to be able to run this.

File: tensorflow/examples/image_retraining/retrain.py [hunk 2]
```diff
@@ -450,7 +452,14 @@ def get_or_create_bottleneck(sess, image_lists, label_name, index, image_dir,
                                 category)
     if not gfile.Exists(image_path):
       tf.logging.fatal('File does not exist %s', image_path)
-    image_data = gfile.FastGFile(image_path, 'rb').read()
+    jpgext = ['jpg', 'jpeg', 'JPG', 'JPEG']
+    if (image_path.split('.')[-1] in jpgext):
```
Comment (rohan100jain) [inline tensorflow/examples/image_retraining/retrain.py]: Use os.path.splitext() instead to figure out extension?

File: tensorflow/examples/image_retraining/retrain.py [hunk 3]
```diff
@@ -450,7 +452,14 @@ def get_or_create_bottleneck(sess, image_lists, label_name, index, image_dir,
                                 category)
     if not gfile.Exists(image_path):
       tf.logging.fatal('File does not exist %s', image_path)
-    image_data = gfile.FastGFile(image_path, 'rb').read()
+    jpgext = ['jpg', 'jpeg', 'JPG', 'JPEG']
```
Comment (rohan100jain) [inline tensorflow/examples/image_retraining/retrain.py]: Make this a separate helper method that takes in file_name and returns back image_data.

File: tensorflow/examples/image_retraining/retrain.py [hunk 4]
```diff
@@ -195,7 +197,7 @@ def create_image_lists(image_dir, testing_percentage, validation_percentage):
     if is_root_dir:
       is_root_dir = False
       continue
-    extensions = ['jpg', 'jpeg', 'JPG', 'JPEG']
+    extensions = ['jpg', 'jpeg', 'JPG', 'JPEG', 'gif', 'png']
```
Comment (rohan100jain) [inline tensorflow/examples/image_retraining/retrain.py]: what about GIF and PNG (caps)

Maybe share this list between here and below? Might also not be a bad idea to make it a dictionary mapping extension to color depth (https://en.wikipedia.org/wiki/Color_depth). That way you can handle the conversion more generally...

**Review by rohan100jain — CHANGES_REQUESTED** [review]

File: tensorflow/examples/image_retraining/retrain.py (hunk 1 above)
Comment (elviswf) [inline tensorflow/examples/image_retraining/retrain.py]: Sure. Thank you!

File: tensorflow/examples/image_retraining/retrain.py (hunk 4 above)
Comment (elviswf) [inline tensorflow/examples/image_retraining/retrain.py]: I have tested these formats. So it goes like:
`extensions = ['jpg', 'jpeg', 'JPG', 'JPEG', 'gif', 'GIF', 'png', 'PNG', 'bmp', 'BMP']`

File: tensorflow/examples/image_retraining/retrain.py (hunk 3 above)
Comment (elviswf) [inline tensorflow/examples/image_retraining/retrain.py]: Sure. I will try it.

File: tensorflow/examples/image_retraining/retrain.py (hunk 2 above)
Comment (elviswf) [inline tensorflow/examples/image_retraining/retrain.py]: os.path.splitext()  returns '.jpg', so it has to be set [1:] to return 'jpg' like:
`os.path.splitext(image_path)[1][1:]` 
I think `image_path.split('.')[-1]` might be  clear enough.

Comment (elviswf) [inline tensorflow/examples/image_retraining/retrain.py]: @rohan100jain @petewarden I have updated the changes. Please check it. Thank you!

**Comment by elviswf** [comment]
@petewarden I have updated the code according to the reviews. Thank you!

File: tensorflow/examples/image_retraining/retrain.py [hunk 5]
```diff
@@ -464,6 +469,29 @@ def get_or_create_bottleneck(sess, image_lists, label_name, index, image_dir,
   return bottleneck_values
 
 
+def read_image2RGBbytes(image_path):
+  """Read the image into bytes stream.
```
Comment (drpngx) [inline tensorflow/examples/image_retraining/retrain.py]: Minor nit: Reads (as per PEP8). Also, "into a bytes stream".

File: tensorflow/examples/image_retraining/retrain.py [hunk 6]
```diff
@@ -464,6 +469,29 @@ def get_or_create_bottleneck(sess, image_lists, label_name, index, image_dir,
...
+
+   For jpg/jpeg, gfile.FastGFile is enough.
```
Comment (drpngx) [inline tensorflow/examples/image_retraining/retrain.py]: Minor nit, this line and the next: remove one leading space.

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by elviswf** [comment]
` Log @: /workspace/pip_test/tests/logs/tensorflow/examples/image_retraining/retrain_test.py.log

============== BEGINS failure log content ==============
Traceback (most recent call last):
  File "/workspace/pip_test/tests/retrain_test.py", line 23, in <module>
    from tensorflow.examples.image_retraining import retrain
  File "/workspace/pip_test/venv/local/lib/python2.7/site-packages/tensorflow/examples/image_retraining/retrain.py", line 81, in <module>
    from PIL import Image
ImportError: No module named PIL

============== ENDS failure log content ==============`

It shows `No module named PIL`. I wonder if there is a way to avoid using PIL package.

**Comment by vrv** [comment]
@petewarden @gunan Should we install PIL as part of our CI infrastructure and add this as a dependency?

Part of me thinks that, serving as an example, this is not going to do everything that people want to do, so perhaps we leave a comment instead suggesting other ways to read images instead.

**Comment by gunan** [comment]
I am not a fan of adding the dependency, as I think everyone will have their own favorite ways of reading images. And PIL is not a real requirement to core TF, so I don't think it is right to brand it like a requirement to TF.

As for installing it to ci, sure.
But that means it will be installed to TF docker images.

**Comment by vrv** [comment]
Okay, closing for now but discussion can continue here if needed.

**Comment by elviswf** [comment]
Actually I fully agree that adding this dependency is not necessary for now. I just used it for a deep learning competition with png, gif data and wanted to shared a way to handle it directly. 
 I will go on with excellent TF. Thank you all!

**Comment by passerbythesun** [comment]
I have the same problem, so this piece code is useful for me.

**Pull request closed without merging**